File: portals/backend/src/db/models/destination.py
from sqlalchemy import Column, Integer, String, Float, Text, ForeignKey, JSON
from sqlalchemy.orm import relationship
from geoalchemy2 import Geometry
from src.db.base_class import Base

class Destination(Base):
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False, index=True)
    description = Column(Text)
    region = Column(String)
    zone = Column(String)
    # Keeping raw floats for easy JSON serialization
    latitude = Column(Float)
    longitude = Column(Float)
    # PostGIS geometry point for spatial queries
    location = Column(Geometry(geometry_type='POINT', srid=4326))
    category = Column(String)
    images = Column(JSON, default=list) # Store list of URLs
    status = Column(String, default="ACTIVE")
    verification_status = Column(String, default="VERIFIED")

File: mobile/lib/data/models/destination.dart
class Destination {
  final String id;
  final String name;
  final String description;
  final String region;
  final String zone;
  final String? woreda;
  final double latitude;
  final double longitude;
  final String category;
  final int? recommendedDurationDays;
  final String verificationStatus;
  final List<String> images;

  const Destination({
    required this.id,
    required this.name,
    this.description = '',
    this.region = '',
    this.zone = '',
    this.woreda,
    required this.latitude,
    required this.longitude,
    this.category = 'destination',
    this.recommendedDurationDays,
    this.verificationStatus = 'VERIFIED',
    this.images = const [],
  });

  String? get coverImage => images.isNotEmpty ? images.first : null;

  Destination copyWith({List<String>? images}) {
    return Destination(
      id: id,
      name: name,
      description: description,
      region: region,
      zone: zone,
      woreda: woreda,
      latitude: latitude,
      longitude: longitude,
      category: category,
      recommendedDurationDays: recommendedDurationDays,
      verificationStatus: verificationStatus,
      images: images ?? this.images,
    );
  }

  factory Destination.fromJson(Map<String, dynamic> json) {
    final imageList = (json['images'] as List<dynamic>?)
            ?.map((e) => e.toString())
            .where((e) => e.isNotEmpty)
            .toList() ??
        const <String>[];
    return Destination(
      id: json['id'].toString(),
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      region: json['region'] as String? ?? '',
      zone: json['zone'] as String? ?? '',
      woreda: json['woreda'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0,
      category: json['category'] as String? ?? 'destination',
      recommendedDurationDays: json['recommended_duration_days'] as int?,
      verificationStatus: json['verification_status'] as String? ?? 'VERIFIED',
      images: imageList,
    );
  }
}
